# Create ISA-API Investigation from Study Design configuration

In this notebook I will show you how you can use a study design configuration is JSON format to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.

Or study design configuration consists of:
* two study arms (treatment and control)
* blood and liver samples are taken during the treatment phase
* blood samples are used to perform Metabolite Profiling with two assays: mass spectrometry and nuclear magnetic resonance (NMR).

## 1. Setup

Let's import all the required libraries

In [1]:
from time import time
import os
import json

## ISA-API related imports
from isatools.model import Investigation, Study

## ISA-API create mode related imports
from isatools.create.models import StudyDesign
from isatools.create.connectors import generate_study_design_from_config

# serializer from ISA Investigation to JSON
from isatools.isajson import ISAJSONEncoder

# ISA-Tab serialisation
from isatools import isatab

[{'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'targeted metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'untargeted metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date

/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/create/models.py:972: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  _kwargs = {key: val for key, val in kwargs.items() if key is not 'qc_sample_type'}


## 2. Load the Study Design JSON configuration

First of all we load the study design configurator

In [2]:
with open(os.path.abspath(os.path.join("config", "study-design-with-two-arms-old.json")), "r") as config_file:
    study_design_config = json.load(config_file)
study_design_config

{'type': 'here you can describe your type of design',
 'elements': [{'id': '#element/screen',
   'name': 'screen',
   'type': 'screen',
   'duration': 7,
   'durationUnit': 'days'},
  {'id': '#element/first_treatment',
   'name': 'first treatment',
   'type': 'chemical intervention',
   'agent': 'test drug',
   'intensity': 10,
   'intensityUnit': 'mg/day',
   'duration': 30,
   'durationUnit': 'days'},
  {'id': '#element/surgical_treatment',
   'name': 'surgical treatment',
   'type': 'surgical intervention',
   'agent': 'surgery'},
  {'id': '#element/follow-up',
   'name': 'follow-up',
   'type': 'follow-up',
   'duration': 3,
   'durationUnit': 'months'},
  {'id': '#element/control_treatment',
   'name': 'control treatment',
   'type': 'chemical intervention',
   'agent': 'placebo',
   'intensity': 10,
   'intensityUnit': 'mg/day',
   'duration': 30,
   'durationUnit': 'days'}],
 'events': [{'id': '#event/sampling_0',
   'action': 'sampling',
   'input': 'subject',
   'output': 'liv

## 3. Generate the ISA Study Design from the JSON configuration
To perform the conversion we just need to use the function `generate_isa_study_design_from_datascriptor_config` (name possibly subject to change, should we drop the "isa" and "datascriptor" qualifiers?)

In [3]:
study_design = generate_study_design_from_config(study_design_config)
assert isinstance(study_design, StudyDesign)

## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation

The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object.

In [4]:
start = time()
study = study_design.generate_isa_study()
end = time()
print('The generation of the study design took {:.2f} s.'.format(end - start))
assert isinstance(study, Study)
investigation = Investigation(studies=[study])

isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnota

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnn

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite p

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite p

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite p

); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometr

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnn

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnn

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=NMR spectroscopy
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnn

isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite prof

); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometr

);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
); {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']};
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite profiling
    term_source=
    term_accession=
    comments=0 Comment objects
); OntologyAnnotation(
    term=mass spectrometry
    term_source=
    term_accession=
    comments=0 Comment objects
);
isa_objects_factory: Assay conf. found: OntologyAnnotation(
    term=metabolite p

## 5. Serialize and save the JSON representation of the generated ISA Investigation

In [5]:
start = time()
inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))
end = time()
print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))

The JSON serialisation of the ISA investigation took 0.35 s.


In [6]:
directory = os.path.abspath(os.path.join('output'))
if not os.path.exists(directory):
    os.makedirs(directory)
with open(os.path.abspath(os.path.join('output','isa-investigation-2-arms-nmr-ms.json')), 'w') as out_fp:
    json.dump(json.loads(inv_json), out_fp)

## 6. Dump the ISA Investigation to ISA-Tab

In [7]:
start = time()
isatab.dump(investigation, os.path.abspath(os.path.join('output')))
end = time()
print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))

2020-06-23 14:56:24,380 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 64 paths!
2020-06-23 14:56:24,387 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 64 paths
2020-06-23 14:56:24,390 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 64 rows
2020-06-23 14:56:28,011 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 384 paths!
2020-06-23 14:56:33,490 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 384 paths!
2020-06-23 14:56:35,819 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 160 paths!
2020-06-23 14:56:40,629 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 480 paths!
2020-06-23 14:56:47,587 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 480 paths!


The Tab serialisation of the ISA investigation took 23.69 s.


To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump

In [8]:
dataframes = isatab.dump_tables_to_dataframes(investigation)

2020-06-23 14:56:48,097 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 64 paths!
2020-06-23 14:56:48,106 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 64 paths
2020-06-23 14:56:48,111 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 64 rows
2020-06-23 14:56:52,110 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 384 paths!
2020-06-23 14:56:57,621 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 384 paths!
2020-06-23 14:57:00,167 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 160 paths!
2020-06-23 14:57:05,133 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 480 paths!
2020-06-23 14:57:12,443 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 480 paths!


In [9]:
len(dataframes)

6

## 7. Check the correctness of the ISA-Tab DataFrames 

We have 1 study file and 5 assay files:
* 1 assay file for Mass. Spec. (treatmennt arm, third epoch: surgery)
* 2 assay files for Mass. Spec. (both arms, fourth epoch: follow-up)
* 2 assay files for NMR (both arms, fourth epoch: follow-up)

In [10]:
for key in dataframes.keys():
    display(key)

's_study_01.txt'

'a_CELL_treatment_3_ASSAY_GRAPH_001_metabolite-profiling_NMR-spectroscopy.txt'

'a_CELL_control_3_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'

'a_CELL_treatment_3_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'

'a_CELL_treatment_2_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'

'a_CELL_control_3_ASSAY_GRAPH_001_metabolite-profiling_NMR-spectroscopy.txt'

We have 8 subjects in the control arm and 24 samples have been collected (3 blood samples per subject during the follow-up epoch)

We have 10 subjects in the control arm and 40 samples have been collected (3 blood samples per subject during the follow-up epoch and 1 liver sample per subject during the surgery epoch)

In [11]:
study_frame = dataframes['s_study_01.txt']
count_control_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'control' in el)])
count_treatment_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'treatment' in el)])
print("There are {} samples in the control arm (i.e. group)".format(count_control_samples))
print("There are {} samples in the treatment arm (i.e. group)".format(count_treatment_samples))

There are 24 samples in the control arm (i.e. group)
There are 40 samples in the treatment arm (i.e. group)


Each control samples is fractioned and 2 labelled extracts are produced (i.e. biological replicates).

```
[
    "labelling",
    {
        "#replicates": 2
    }
]
```

There are 2 possible combinations of Mass Spec Assay specified in our configuration template and 2 techincal replicate are produced for each combinations:

```
[
    "mass spectrometry",
    {
        "#replicates": 2,
        "instrument": [
            "Agilent QTQF 6510"
        ],
        "injection_mode": [
            "FIA",
            "LC"
        ],
        "acquisition_mode": [
            "positive mode"
        ]
    }
]
```

Two output raw spectral files are produced as the result of reach run

```
[
    "raw spectral data file",
    [
        {
            "node_type": "data file",
            "size": 2,
            "is_input_to_next_protocols": false
        }
    ]
]
```

As a total we expect

$$ N_{rows} = (N_{subjects} \times N_{samples}) \times (N_{biorepl} \times N_{combinations} \times N_{techrepl}) = (8 \times 3) \times (2 \times 2 \times 2 \times 2) = 24 \times 16 = 384 $$

Which we can verify for the mass spectrometry assay file of the control group.

In [12]:
dataframes['a_CELL_treatment_3_ASSAY_GRAPH_001_metabolite-profiling_NMR-spectroscopy.txt']

Sample Name Protocol REF Performer  \
0    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
1    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
2    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
3    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
4    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
..                                                 ...          ...       ...   
475  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
476  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
477  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
478  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   
479  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   

      Extract Name Characteristics[extract type]    Protocol REF.1  \
0    extract_27-18                        pellet  nmr spectroscopy   
1     extract_27-1                   supernatant  nmr spectroscopy   
2    extract_27-18                        pellet  nmr spectroscopy   
3     extract_27-1                   supernatant  nmr spectroscopy   
4    extract_27-18                        pellet  nmr spectroscopy   
..             ...                           ...               ...   
475  extract_11-18                        pellet  nmr spectroscopy   
476  extract_11-18                        pellet  nmr spectroscopy   
477   extract_11-1                   supernatant  nmr spectroscopy   
478  extract_11-18                        pellet  nmr spectroscopy   
479   extract_11-1                   supernatant  nmr spectroscopy   

    Parameter Value[instrument] Parameter Value[pulse_sequence]  \
0         Bruker AvanceII 1 GHz                            CPMG   
1         Bruker AvanceII 1 GHz                            CPMG   
2         Bruker AvanceII 1 GHz                            CPMG   
3         Bruker AvanceII 1 GHz                       watergate   
4         Bruker AvanceII 1 GHz                       watergate   
..                          ...                             ...   
475       Bruker AvanceII 1 GHz                            CPMG   
476       Bruker AvanceII 1 GHz                            CPMG   
477       Bruker AvanceII 1 GHz                       watergate   
478       Bruker AvanceII 1 GHz                            CPMG   
479       Bruker AvanceII 1 GHz                            CPMG   

    Parameter Value[acquisition_mode] Performer.1  \
0                      2D 13C-13C NMR    Ellipsis   
1                          1D 13C NMR    Ellipsis   
2                          1D 13C NMR    Ellipsis   
3                          1D 13C NMR    Ellipsis   
4                          1D 13C NMR    Ellipsis   
..                                ...         ...   
475                    2D 13C-13C NMR    Ellipsis   
476                        1D 13C NMR    Ellipsis   
477                    2D 13C-13C NMR    Ellipsis   
478                    2D 13C-13C NMR    Ellipsis   
479                    2D 13C-13C NMR    Ellipsis   

    Free Induction Decay Data File  
0     raw spectral data file_27-26  
1      raw spectral data file_27-3  
2     raw spectral data file_27-20  
3     raw spectral data file_27-11  
4     raw spectral data file_27-30  
..                             ...  
475   raw spectral data file_11-24  
476   raw spectral data file_11-22  
477    raw spectral data file_11-7  
478   raw spectral data file_11-26  
479   raw spectral data file_11-15  

[480 rows x 11 columns]